File: subscription-tracker-app/src/app/services/subscription.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Subscription, CreateSubscription, UpdateSubscription, SubscriptionQuery } from '../models/subscription.model';

@Injectable({
  providedIn: 'root'
})
export class SubscriptionService {
  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/subscriptions`;

  // GET: كل اشتراكات مستخدم معين، مع دعم البحث والفلترة والترتيب
  getAllForUser(userId: number, query?: SubscriptionQuery): Observable<Subscription[]> {
    let params = new HttpParams();

    if (query?.search) params = params.set('search', query.search);
    if (query?.status !== undefined) params = params.set('status', query.status);
    if (query?.billingCycle !== undefined) params = params.set('billingCycle', query.billingCycle);
    if (query?.categoryId !== undefined) params = params.set('categoryId', query.categoryId);
    if (query?.tagId !== undefined) params = params.set('tagId', query.tagId);
    if (query?.onlyFavorites !== undefined) params = params.set('onlyFavorites', query.onlyFavorites);
    if (query?.renewalFrom) params = params.set('renewalFrom', query.renewalFrom);
    if (query?.renewalTo) params = params.set('renewalTo', query.renewalTo);
    if (query?.sortBy) params = params.set('sortBy', query.sortBy);
    if (query?.sortDescending !== undefined) params = params.set('sortDescending', query.sortDescending);

    return this.http.get<Subscription[]>(`${this.baseUrl}/user/${userId}`, { params });
  }

  // GET: اشتراك واحد بالتفصيل
  getById(id: number): Observable<Subscription> {
    return this.http.get<Subscription>(`${this.baseUrl}/${id}`);
  }

  // POST: إضافة اشتراك جديد
  create(dto: CreateSubscription): Observable<Subscription> {
    return this.http.post<Subscription>(this.baseUrl, dto);
  }

  // PUT: تعديل اشتراك موجود
  update(id: number, dto: UpdateSubscription): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${id}`, dto);
  }

  // DELETE: حذف اشتراك
  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // POST: عمل نسخة من اشتراك موجود
  duplicate(id: number): Observable<Subscription> {
    return this.http.post<Subscription>(`${this.baseUrl}/${id}/duplicate`, {});
  }

  // بيقلب حالة المفضلة - بيبني الـ payload الكامل من الاشتراك نفسه (مفيش Endpoint منفصل للـ toggle بس)
  toggleFavorite(sub: Subscription): Observable<void> {
    const dto: UpdateSubscription = {
      name: sub.name,
      description: sub.description,
      price: sub.price,
      currency: sub.currency,
      billingCycle: sub.billingCycle,
      startDate: sub.startDate,
      nextRenewalDate: sub.nextRenewalDate,
      autoRenew: sub.autoRenew,
      isFavorite: !sub.isFavorite,
      websiteUrl: sub.websiteUrl,
      notes: sub.notes,
      categoryId: sub.category?.id ?? null,
      paymentMethodId: sub.paymentMethod?.id ?? null,
      status: sub.status,
      tagIds: sub.tags.map((t) => t.id),
      icon: sub.icon
    };
    return this.update(sub.id, dto);
  }

  // GET: إجمالي المصروف الشهري
  getMonthlyTotal(userId: number): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/user/${userId}/monthly-total`);
  }
}

File: CleanArch-updated/src/SubscriptionTracker.Application/Services/SubscriptionService.cs
using Microsoft.EntityFrameworkCore;
using SubscriptionTracker.Application.DTOs;
using SubscriptionTracker.Application.Interfaces.Repositories;
using SubscriptionTracker.Application.Interfaces.Services;
using SubscriptionTracker.Application.Mapping;
using SubscriptionTracker.Domain.Common;
using SubscriptionTracker.Domain.Entities;
using SubscriptionTracker.Domain.Enums;

namespace SubscriptionTracker.Application.Services
{
    public class SubscriptionService : ISubscriptionService
    {
        private readonly IUnitOfWork _unitOfWork;

        public SubscriptionService(IUnitOfWork unitOfWork)
        {
            _unitOfWork = unitOfWork;
        }

        public async Task<List<SubscriptionDto>> GetAllForUserAsync(int userId, SubscriptionQueryOptions options)
        {
            var query = _unitOfWork.Subscriptions.QueryWithDetails()
                .Where(s => s.UserId == userId);

            if (!string.IsNullOrWhiteSpace(options.Search))
            {
                var search = options.Search.Trim().ToLower();
                query = query.Where(s => s.Name.ToLower().Contains(search));
            }

            if (options.Status.HasValue)
                query = query.Where(s => s.Status == options.Status.Value);

            if (options.BillingCycle.HasValue)
                query = query.Where(s => s.BillingCycle == options.BillingCycle.Value);

            if (options.CategoryId.HasValue)
                query = query.Where(s => s.CategoryId == options.CategoryId.Value);

            if (options.TagId.HasValue)
                query = query.Where(s => s.Tags.Any(t => t.Id == options.TagId.Value));

            if (options.OnlyFavorites == true)
                query = query.Where(s => s.IsFavorite);

            if (options.RenewalFrom.HasValue)
                query = query.Where(s => s.NextRenewalDate >= options.RenewalFrom.Value);

            if (options.RenewalTo.HasValue)
                query = query.Where(s => s.NextRenewalDate <= options.RenewalTo.Value);

            query = options.SortBy switch
            {
                SubscriptionSortBy.Cost => options.SortDescending
                    ? query.OrderByDescending(s => s.Price)
                    : query.OrderBy(s => s.Price),
                SubscriptionSortBy.Name => options.SortDescending
                    ? query.OrderByDescending(s => s.Name)
                    : query.OrderBy(s => s.Name),
                _ => options.SortDescending
                    ? query.OrderByDescending(s => s.NextRenewalDate)
                    : query.OrderBy(s => s.NextRenewalDate)
            };

            var subscriptions = await query.ToListAsync();
            return subscriptions.ToDtoList();
        }

        public async Task<SubscriptionDto?> GetByIdAsync(int id)
        {
            var subscription = await _unitOfWork.Subscriptions.GetByIdWithDetailsAsync(id);
            return subscription?.ToDto();
        }

        public async Task<SubscriptionDto> CreateAsync(CreateSubscriptionDto dto)
        {
            var subscription = new Subscription
            {
                Name = dto.Name,
                Description = dto.Description,
                Price = dto.Price,
                Currency = dto.Currency,
                BillingCycle = dto.BillingCycle,
                StartDate = dto.StartDate,
                NextRenewalDate = dto.NextRenewalDate,
                AutoRenew = dto.AutoRenew,
                WebsiteUrl = dto.WebsiteUrl,
                Notes = dto.Notes,
                IsFavorite = dto.IsFavorite,
                Icon = dto.Icon,
                CategoryId = dto.CategoryId,
                PaymentMethodId = dto.PaymentMethodId,
                UserId = dto.UserId,
                Status = SubscriptionStatus.Active,
                CreatedAt = DateTime.UtcNow
            };

            if (dto.TagIds is { Count: > 0 })
            {
                subscription.Tags = await _unitOfWork.Tags.Query()
                    .Where(t => dto.TagIds.Contains(t.Id))
                    .ToListAsync();
            }

            await _unitOfWork.Subscriptions.AddAsync(subscription);
            await _unitOfWork.SaveChangesAsync();

            // بعد الحفظ، بنجيب النسخة كاملة مع كل العلاقات عشان الـ DTO يرجع متكامل
            // (! آمن هنا: لسه فاكرين الـ Id من subscription.Id اللي اتحفظ فوق - مستحيل يرجع null)
            var created = await _unitOfWork.Subscriptions.GetByIdWithDetailsAsync(subscription.Id);
            return created!.ToDto();
        }

        public async Task<bool> UpdateAsync(int id, UpdateSubscriptionDto dto)
        {
            var subscription = await _unitOfWork.Subscriptions.GetByIdWithDetailsAsync(id);
            if (subscription is null) return false;

            subscription.Name = dto.Name;
            subscription.Description = dto.Description;
            subscription.Price = dto.Price;
            subscription.Currency = dto.Currency;
            subscription.BillingCycle = dto.BillingCycle;
            subscription.StartDate = dto.StartDate;
            subscription.NextRenewalDate = dto.NextRenewalDate;
            subscription.AutoRenew = dto.AutoRenew;
            subscription.WebsiteUrl = dto.WebsiteUrl;
            subscription.Notes = dto.Notes;
            subscription.IsFavorite = dto.IsFavorite;
            subscription.Icon = dto.Icon;
            subscription.CategoryId = dto.CategoryId;
            subscription.PaymentMethodId = dto.PaymentMethodId;
            subscription.Status = dto.Status;

            // بنستبدل التاجز بالكامل بالقائمة الجديدة (أبسط من مقارنة الفروقات)
            if (dto.TagIds is not null)
            {
                subscription.Tags = await _unitOfWork.Tags.Query()
                    .Where(t => dto.TagIds.Contains(t.Id))
                    .ToListAsync();
            }

            await _unitOfWork.SaveChangesAsync();
            return true;
        }

        public async Task<bool> DeleteAsync(int id)
        {
            var subscription = await _unitOfWork.Subscriptions.GetByIdAsync(id);
            if (subscription is null) return false;

            _unitOfWork.Subscriptions.Remove(subscription);
            await _unitOfWork.SaveChangesAsync();
            return true;
        }

        public async Task<SubscriptionDto?> DuplicateAsync(int id)
        {
            var original = await _unitOfWork.Subscriptions.GetByIdWithDetailsAsync(id);
            if (original is null) return null;

            var copy = new Subscription
            {
                // بنضيف "(نسخة)" لاسم الاشتراك عشان يبقى واضح إنه مكرر
                Name = $"{original.Name} (نسخة)",
                Description = original.Description,
                Price = original.Price,
                Currency = original.Currency,
                BillingCycle = original.BillingCycle,
                StartDate = original.StartDate,
                NextRenewalDate = original.NextRenewalDate,
                AutoRenew = original.AutoRenew,
                WebsiteUrl = original.WebsiteUrl,
                Notes = original.Notes,
                IsFavorite = false, // النسخة مبتورثش حالة المفضلة
                Icon = original.Icon,
                CategoryId = original.CategoryId,
                PaymentMethodId = original.PaymentMethodId,
                UserId = original.UserId,
                Status = SubscriptionStatus.Active,
                CreatedAt = DateTime.UtcNow,
                Tags = original.Tags.ToList()
            };

            await _unitOfWork.Subscriptions.AddAsync(copy);
            await _unitOfWork.SaveChangesAsync();

            var created = await _unitOfWork.Subscriptions.GetByIdWithDetailsAsync(copy.Id);
            return created!.ToDto();
        }

        public async Task<decimal> GetMonthlyTotalAsync(int userId)
        {
            var subscriptions = await _unitOfWork.Subscriptions.Query()
                .Where(s => s.UserId == userId && s.Status == SubscriptionStatus.Active)
                .ToListAsync();

            return subscriptions.Sum(s => BillingCycleHelper.ToMonthlyEquivalent(s.Price, s.BillingCycle));
        }
    }
}
